src/utils/useLeads.tsx

```diff
@@ -0,0 +1,88 @@
+import { useEffect, useState } from 'react'
+
+import { Lead } from 'types/Lead'
+import { useConnection } from './useConnection'
+import { usePrevious } from '@apideck/components'
+import useSWR from 'swr'
+
+export const useLeads = () => {
+  const [cursor, setCursor] = useState(null)
+  const { connection } = useConnection()
+  const serviceId = connection?.service_id || ''
+  const prevServiceId = usePrevious(serviceId)
+  const prevCursor = usePrevious(cursor)
+
+  const fetcher = async (url: string) => {
+    const response = await fetch(url)
+    return await response.json()
+  }
+
+  const cursorParams =
+    cursor && (!prevServiceId || prevServiceId === serviceId) ? `&cursor=${cursor}` : ''
+  const getLeadsUrl = `/api/crm/leads/get?serviceId=${serviceId}${cursorParams}`
+  const { data, error, revalidate } = useSWR(getLeadsUrl, fetcher)
+
+  useEffect(() => {
+    if (prevServiceId && prevServiceId !== serviceId) {
+      setCursor(null)
+    }
+  }, [serviceId, prevServiceId])
+
+  const createLead = async (values: Lead) => {
+    const response = await fetch(`/api/crm/leads/post?serviceId=${serviceId}`, {
+      method: 'POST',
+      body: JSON.stringify(values)
+    })
+    return response.json()
+  }
+
+  const updateLead = async (id: string, values: Lead) => {
+    const response = await fetch(`/api/crm/leads/patch?serviceId=${serviceId}`, {
+      method: 'PATCH',
+      body: JSON.stringify({ id, ...values })
+    })
+    return response.json()
+  }
+
+  const deleteLead = async (id: string) => {
+    const response = await fetch(`/api/crm/leads/delete?serviceId=${serviceId}`, {
+      method: 'DELETE',
+      body: JSON.stringify({ id })
+    })
+    return response.json()
+  }
+
+  const nextPage = () => {
+    const nextCursor = data?.meta?.cursors?.next
+
+    if (nextCursor) {
+      setCursor(nextCursor)
+    }
+  }
+
+  const prevPage = () => {
+    const prevCursor = data?.meta?.cursors?.previous
+    setCursor(prevCursor || null)
+  }
+
+  useEffect(() => {
+    if (prevCursor && prevCursor !== cursor) {
+      revalidate()
+    }
+  }, [cursor, prevCursor, revalidate])
+
+  return {
+    leads: data,
+    isLoading: !error && !data,
+    isError: data?.error || error,
+    hasNextPage: data?.meta?.cursors?.next,
+    currentPage: data?.meta?.cursors?.current,
+    hasPrevPage: data?.meta?.cursors?.previous,
```

Comment:
Did manage to fix it.

To show the button:
```
 {(hasPrevPage || (!hasPrevPage && currentPage)) && (
      <Button onClick={() => prevPage()} text="Previous" isLoading={isLoading} />
 )}
```

To go the the previous page:
```
const prevPage = () => {
    const prevCursor = data?.meta?.cursors?.previous
    setCursor(prevCursor)
}
```

The effect that triggers a new fetch:
```
useEffect(() => {
    if (prevCursor && prevCursor !== cursor) {
      revalidate()
    }
 }, [cursor, prevCursor, revalidate])
  ```

So if `cursor` is `null` it will remove the cursor params when fetching so it gets the first page. Tested this out and works fine now. Do you see any potential issues with this?

Reply:
No looks good :smile: Just wondering if it would be nicer if the API returned a cursor for the first page instead of `null`. I think it might be possible.

You could also use the `links`, don't have to change it now but it might be slightly easier to work with.

Reply:
yea I think it could indeed be a bit nicer if the API could return that. Then I only have to check if the previous cursor is present. Thought that wasn't possible. Maybe you can add an issue for that?

The links are nice, but I'm going to take a look at the SDK now so not going to need that. Not sure if pagination is already supported by the SDK but I'll check that with Geert ;)